core/src/test/java/com/graphhopper/routing/weighting/TurnWeightingTest.java

```diff
@@ -0,0 +1,62 @@
+package com.graphhopper.routing.weighting;
+
+import com.graphhopper.routing.util.CarFlagEncoder;
+import com.graphhopper.routing.util.EncodingManager;
+import com.graphhopper.routing.util.FlagEncoder;
+import com.graphhopper.storage.Graph;
+import com.graphhopper.storage.GraphBuilder;
+import com.graphhopper.storage.TurnCostExtension;
+import com.graphhopper.util.EdgeIteratorState;
+import org.junit.Before;
+import org.junit.Test;
+
+import static com.graphhopper.util.GHUtility.getEdge;
+import static org.junit.Assert.assertEquals;
+
+public class TurnWeightingTest {
+
+    private Graph graph;
+    private FlagEncoder encoder;
+    private EncodingManager encodingManager;
+    private Weighting weighting;
+    private TurnCostExtension turnCostExt;
+
+    @Before
+    public void setup() {
+        encoder = new CarFlagEncoder(5, 5, 10);
+        encodingManager = EncodingManager.create(encoder);
+        graph = new GraphBuilder(encodingManager).create();
+        weighting = new FastestWeighting(encoder);
+        turnCostExt = (TurnCostExtension) graph.getExtension();
+    }
+
+    @Test
+    public void calcWeightAndTime_withTurnCosts() {
+        graph.edge(0, 1, 100, true);
+        EdgeIteratorState edge = graph.edge(1, 2, 100, true);
+        // turn costs are given in seconds
+        addTurnCost(0, 1, 2, 5);
+        TurnWeighting turnWeighting = new TurnWeighting(weighting, turnCostExt);
+        assertEquals(6 + 5, turnWeighting.calcWeight(edge, false, 0), 1.e-6);
+        assertEquals(6000 + 5000, turnWeighting.calcMillis(edge, false, 0), 1.e-6);
+    }
+
+    @Test
+    public void calcWeightAndTime_withTurnCosts_shortest() {
+        graph.edge(0, 1, 100, true);
+        EdgeIteratorState edge = graph.edge(1, 2, 100, true);
+        // turn costs are given in seconds
+        addTurnCost(0, 1, 2, 5);
+        TurnWeighting turnWeighting = new TurnWeighting(new ShortestWeighting(encoder), turnCostExt);
+        // todo: for the shortest weighting turn costs cannot be interpreted as seconds ? at least when they are added
+        // to the weight ? how much should they contribute ?
+//        assertEquals(105, turnWeighting.calcWeight(edge, false, 0), 1.e-6);
```

Comment:
Any thoughts here ?

Reply:
What do you think about deprecating shortest weighting in general? It does not make sense in practise. Most people say they need "shortest" for car (hear this often for taxis as they calculate the cost based on the distance), but what they really mean is a "fastest" cost calculation with a tight distance weight.

Reply:
Ok I see. But not sure if it should be really deprecated, because someone might find it still useful. We could emphasize somewhere that the ShortestFastestWeighting is typically what most people are looking for. Anyway the question how turn costs should contribute to weightings that are not purely based on time remains also for the ShortestFastestWeighting. In ShortestFastestWeighting this is controlled by the time/distanceFactors.

Reply:
For now IMO the turn costs for ShortestWeighting must be 0 as there are no ways to introduce a timely slow down or priority stuff.

In ShortestFastestWeighting the turn cost adds to the time (and gets multiplied by the timeFactor)

Reply:
Yes, I agree as long as we do not allow giving less priority to routes with many turns etc. the shortest weighting should pay attention to turn restrictions, but the turn costs (assuming they are turn *times*) should not be included. However to make this work we would in fact have to do this:

>  Or even introduce a separate method calcTurnMillis like we have for normal weighting ... but are there really cases where a priority to avoid a turn is different to the turn time?

because right now the TurnWeighting simply adds the turn costs without knowing what it's underlying Weighting is.

Reply:
Hmmh, indeed. This is ugly and one reason more (for me) to remove ShortestWeighting :D

Reply:
see #1590